### PR TITLE
fix(redux): checkout reducer not overwriting address state

### PR DESCRIPTION
**Pull request by anapim**
## Description

This fixes an issue where the properties `state.code` and `state.name` in the shipping/billing address of a checkout order would not be cleared in the store if their value was `undefined`.

### Dependencies

None.

## Checklist

- [x] The commit message follows our guidelines
- [x] Tests for the respective changes have been added
- [ ] The code is commented, particularly in hard-to-understand areas
- [x] The labels and/or milestones were added

## Disclaimer

By sending us your contributions, you are agreeing that your contribution is made subject to the terms of our [Contributor Ownership Statement](https://github.com/Farfetch/.github/blob/master/COS.md)

### CODE REVIEW

File: packages/redux/src/checkout/reducer.ts [hunk 1]
```diff
@@ -172,20 +172,70 @@ const convertCheckoutOrder = (
...
+      !!get(
+        mergedState,
+        `checkoutOrders[${result}].billingAddress.state.name`,
+      )) &&
+    (!get(entities, `checkoutOrders[${result}].billingAddress.state.code`) ||
+      !get(entities, `checkoutOrders[${result}].billingAddress.state.name`));
+
+  let returnedState = mergedState;
+
+  if (shouldResetShippingState || shouldResetBillingState) {
```
Comment (boliveira) [inline packages/redux/src/checkout/reducer.ts:201]: This will replace both addresses if only one of them has changed, is that the intention here?

File: packages/redux/src/checkout/reducer.ts [hunk 2]
```diff
@@ -172,20 +172,70 @@ const convertCheckoutOrder = (
...
+    returnedState = {
+      ...returnedState,
```
Comment (boliveira) [inline packages/redux/src/checkout/reducer.ts:203]: As stated before, using immer will make these updates a lot easier and less error prone.

File: packages/redux/src/checkout/__tests__/reducer.test.ts [hunk 3]
```diff
@@ -1326,6 +1326,80 @@ describe('checkout reducer', () => {
...
+              },
+              billingAddress: {
+                state: {},
+              },
+            },
+          },
+        };
+
+        expect(
+          // @ts-expect-error
```
Comment (boliveira) [inline packages/redux/src/checkout/__tests__/reducer.test.ts:1364]: Why was this `ts-expect-error` necessary? If there is a valid reason, add the reason in front of it as it let others know why was it necessary without removing the comment.

File: packages/redux/src/checkout/reducer.ts [hunk 4]
```diff
@@ -172,20 +172,70 @@ const convertCheckoutOrder = (
     !!get(mergedState, `checkoutOrders[${result}].clickAndCollect`) &&
     !get(entities, `checkoutOrders[${result}].clickAndCollect`);
 
-  if (!shouldResetClickAndCollect) {
-    return mergedState;
+  const shouldResetShippingState =
+    (!!get(
```
[the remark this answers is not in the document]
Comment (anapim) [reply packages/redux/src/checkout/reducer.ts]: I followed your suggestions, can you review again?

File: packages/redux/src/checkout/reducer.ts [hunk 5]
```diff
@@ -172,20 +173,72 @@ const convertCheckoutOrder = (
...
+        collectpoints: draftState.checkoutOrders[result]?.collectpoints,
         ...entities?.checkoutOrders[result],
-      },
-    },
-  };
+      };
+    }
+
+    if (shouldReplaceShippingState) {
+      draftState.checkoutOrders[result] = {
```
Comment (boliveira) [inline packages/redux/src/checkout/reducer.ts:223]: It is not necessary to reassign the checkout order in draftState since immer will take care of that.
In this case what you wanted is to do something like this:

```js
draftState.checkoutOrders[result]?.shippingAddress.state = entities.checkoutOrders[result]?.shippingAddress?.state;
```

Then, when the call to `produce` finishes, immer will check which properties have changed and will generate a new state object with new references only for those properties that changed.

If, instead, you meant to replace the whole shipping address if `state` has changed, then you will need to adapt to something like this:

```js
draftState.checkoutOrders[result]?.shippingAddress= entities.checkoutOrders[result]?.shippingAddress;
```
So, in summary, you only need to treat the `draftState` as a mutable object and immer will take care of generating the object will new references where necessary.

**Review by andrevechina — APPROVED** [review]

**Review by boliveira — APPROVED** [review]

**Review by edmundofreitas — APPROVED** [review]

**Review by nelsonleite — APPROVED** [review]
